### ISSUE TITLE
`F.scaled_dot_product_attention` returns all-`NaN` rows during evaluation but not training

### ISSUE THREAD

**Issue by amorehead** [pytorch/pytorch#163997]
### 🐛 Describe the bug

When my `nn.Module`'s training mode is set (`self.training == True`), `F.scaled_dot_product_attention` works as intended with `attn_mask`s that contain some rows with all `-inf` values (denoting that these are purely padding rows). However, after upgrading from PyTorch 2.6.0 to 2.8.0, when `self.training == False` (i.e., when I am running a validation epoch with the `nn.Module`), `F.scaled_dot_product_attention` starts returning all-`NaN` rows for positions that (originally) denoted padding rows (instead of returning all-zero rows as expected during training). 

Notably, during training, my batch size `B=2`, whereas during validation `B=1`. I've tried calling `.repeat(2, 1, 1, 1)` to see if the batch size was related to the issue, but I found it was not.

```python
# when `self.training == False`
x = F.scaled_dot_product_attention(
    q,  # Float[B, H, N, C]
    k,  # Float[B, H, N, C]
    v,  # Float[B, H, N, C]
    attn_mask=attn_mask,  # Float[B, 1, N, N], where some rows are all `-inf` (as anticipated)
    dropout_p=0.0,
) # now, the last few (padding) rows of `x` are all `NaN`s
```

### Versions

```
Collecting environment information...
PyTorch version: 2.8.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 19.1.7
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.10.18 | packaged by conda-forge | (main, Jun  4 2025, 14:46:00) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4 Max

Versions of relevant libraries:
[pip3] lovely-numpy==0.2.15
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] pytorch-lightning==2.5.5
[pip3] torch==2.8.0
[pip3] torch-geometric==2.6.1
[pip3] torch_scatter==2.1.2
[pip3] torchdiffeq==0.2.5
[pip3] torchmetrics==1.8.1
[pip3] torchtnt==0.2.4
[pip3] torchvision==0.23.0
[pip3] triton==3.5.0+git1a7b5192
[conda] lovely-numpy                        0.2.15                 pypi_0               pypi
[conda] numpy                               2.2.6                  pypi_0               pypi
[conda] pytorch-lightning                   2.5.5                  pypi_0               pypi
[conda] torch                               2.8.0                  pypi_0               pypi
[conda] torch-geometric                     2.6.1                  pypi_0               pypi
[conda] torch-scatter                       2.1.2                  pypi_0               pypi
[conda] torchdiffeq                         0.2.5                  pypi_0               pypi
[conda] torchmetrics                        1.8.1                  pypi_0               pypi
[conda] torchtnt                            0.2.4                  pypi_0               pypi
[conda] torchvision                         0.23.0                 pypi_0               pypi
[conda] triton                              3.5.0+git1a7b5192      pypi_0               pypi
```
